lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
+type KeyValueMap = {
+    [TKey in Key | CollectionKey]: Value[TKey];
+};
+
+type BaseConnectOptions<TKey extends Key | CollectionKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, Value[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends Key | CollectionKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, Value[TKey]> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
```

Comment:
should `waitForCollectionCallback` be optional?

Reply:
so if an item in a collection is deleted, is the key also deleted?

in other words, we don't have this kind of situation?

`Record<string, Value[TKey] | null>`

Reply:
It is optional underneath `waitForCollectionCallback?: false;` @hayata-suenaga 

From what I can tell this parameter makes sense only for collection keys and that's why Fabio created this union type. Notice that in the first type the key has to be a collection key `key: TKey extends CollectionKey ? TKey : never;` and in second key can be a normal key and collection key.

Reply:
🔖 Stopped here, will be OOO this afternoon and will continue answering the comments at night.

Reply:
Yes, `waitForCollectionCallback` is optional and will work only when using collection keys.

Reply:
so `waitForCollectionCallback` should be `?` here? This is a union. This is saying that when we have collection, we always have to have `waitForCollectionCallback`

Reply:
I feel like this accomplishes what we want to accomplish more accurately?

```
type ConnectionOptions<TKey extends Key | CollectionKey> =
  TKey extends CollectionKey
    ? {
        key: TKey;
        callback?: (
          value: Record<string, Value[TKey]> | null,
          key?: TKey
        ) => void;
        waitForCollectionCallback?: boolean;
      }
    : {
        key: TKey;
        callback?: (value: Value[TKey] | null, key?: TKey) => void;
        waitForCollectionCallback?: false;
      };
```

Reply:
I think I'm happy with the code as it's written currently 👍

Reply:
well the original definition and my suggestion have different effects

- the original definition creates an union type.
- in the original definition, you cannot pass `false` for `waitForCollectionCallback` if the key is a collection key

Reply:
> * in the original definition, you cannot pass `false` for `waitForCollectionCallback` if the key is a collection key

I think you can. The first shape in the union type doesn't match, so it goes to the next. A `CollectionKey` is a `TKey` so it matches the key field. `waitForCollectionCallback` is false, so that matches, and in that case the callback definition also matches.

Does that sound right guys?

Reply:
good observation 👍 

one thing missing here, however, is that the type of `value` argument to the callback is `value: Value[TKey]` vs `Record<string, Value[TKey]>` in that case.

or is it supposed to be `Value[TKey]`? 🤔

Reply:
```ts
// error, "waitForCollectionCallback" can only be true if we
// are using a collection.
Onyx.connect({ key: ONYXKEYS.ACCOUNT, callback: (value) => {}, waitForCollectionCallback: true, });
```

wait is this true? I was assuming this can be both `true` and `false`. why does this parameter exist to begin with then? 😕 

```
// error, "waitForCollectionCallback" can only be true if we
// are using a collection.
```

cc: @fabioh8010 @blazejkustra @neil-marcellini

Reply:
>  @param {Boolean} [mapping.waitForCollectionCallback] If set to true, it will return the entire collection to the callback as a single object

If I understand well, `waitForCollectionCallback` is a parameter which should be used only with collection keys. That's why it errors if we use it with non collection keys. Does it make sense? @fabioh8010

Reply:
> If I understand well, waitForCollectionCallback is a parameter which should be used only with collection keys. That's why it errors if we use it with non collection keys. Does it make sense? @fabioh8010

Yes that is correct @blazejkustra ! Looking at the code, `waitForCollectionCallback: true` is only meant to be used together with a collection key, that's why I added this restriction. cc @hayata-suenaga

Reply:
@hayata-suenaga When using a collection key, if `waitForCollectionCallback` is `true` it will fire one `callback` passing all the objects of that collection when we got one update in that collection.

If it's `false`, it will fire one `callback` for EACH item of the collection.

Reply:
sorry I think I quoted a wrong example code [here](https://github.com/Expensify/react-native-onyx/pull/267/files#r1274194341) 😓 

So in the code below, the callback is fired for each different report?

```ts
Onyx.connect({ key: ONYXKEYS.COLLECTION.REPORT, callback: (value) => {}, // "value" is still Report | null. waitForCollectionCallback: false, });
```

@fabioh8010

Reply:
@hayata-suenaga Yes, when `waitForCollectionCallback` is `false` or `undefined` like your example, it will fire for each Report.